File: exvisit/ast.py
"""AST for the .exv DSL."""
from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional, Tuple


class EdgeKind(str, Enum):
    SYNC = "->"
    ASYNC = "~>"


Bounds = Tuple[int, int, int, int]  # x, y, w, h
LineRange = Tuple[int, int]  # start_line, end_line


@dataclass
class Node:
    name: str
    bounds: Bounds
    src_path: Optional[str] = None
    line_range: Optional[LineRange] = None
    states: List[str] = field(default_factory=list)  # ordered state machine
    ns_path: str = ""  # filled by parser: dotted namespace path this node lives in

    @property
    def fqn(self) -> str:
        return f"{self.ns_path}.{self.name}" if self.ns_path else self.name


@dataclass
class Edge:
    src: str  # fully-qualified dotted name
    dst: str
    kind: EdgeKind = EdgeKind.SYNC


@dataclass
class Namespace:
    level: int
    name: str
    bounds: Bounds
    src_glob: Optional[str] = None
    children: List["Namespace"] = field(default_factory=list)
    nodes: List[Node] = field(default_factory=list)
    path: str = ""  # dotted path

    def iter_nodes(self):
        yield from self.nodes
        for c in self.children:
            yield from c.iter_nodes()

    def iter_namespaces(self):
        yield self
        for c in self.children:
            yield from c.iter_namespaces()


@dataclass
class exvisitDoc:
    root: Namespace
    edges: List[Edge] = field(default_factory=list)

    def all_nodes(self) -> List[Node]:
        return list(self.root.iter_nodes())

    def find_node(self, fqn: str) -> Optional[Node]:
        # accept bare name if unique
        if "." in fqn:
            for n in self.all_nodes():
                if n.fqn == fqn or n.fqn.endswith("." + fqn):
                    return n
            return None
        matches = [n for n in self.all_nodes() if n.name == fqn]
        return matches[0] if len(matches) == 1 else None

File: rust/crates/exvisit-core/src/ast.rs
//! Abstract Syntax Tree — mirrors exvisit_pro/exvisit/ast.py.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum EdgeKind {
    /// `->` static import / direct call dependency. Enforced by `exvisit-lsp`.
    Sync,
    /// `~>` runtime wire (event / callback / signal). Informational only.
    Async,
}

/// Local bounds within the parent namespace. `(x, y, w, h)` in arbitrary units.
pub type Bounds = (i32, i32, i32, i32);

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Node {
    pub name: String,
    pub bounds: Bounds,
    pub src_path: Option<String>,
    pub states: Vec<String>,
    pub ns_path: String,
}

impl Node {
    pub fn fqn(&self) -> String {
        if self.ns_path.is_empty() { self.name.clone() }
        else { format!("{}.{}", self.ns_path, self.name) }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Edge {
    pub src: String,
    pub dst: String,
    pub kind: EdgeKind,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Namespace {
    pub level: u8,
    pub name: String,
    pub bounds: Bounds,
    pub src_glob: Option<String>,
    pub children: Vec<Namespace>,
    pub nodes: Vec<Node>,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct exvisitDoc {
    pub root: Namespace,
    pub edges: Vec<Edge>,
}

